aptos-move/framework/aptos-framework/sources/account/account.move

```diff
@@ -424,6 +443,185 @@ module aptos_framework::account {
+        let public_key_type = keyless_public_key[0];
+        assert!(public_key_type == 3 || public_key_type == 4, std::error::invalid_argument(ENOT_A_KEYLESS_PUBLIC_KEY));
+
+        let addr = signer::address_of(account);
+        let account_resource = &mut Account[addr];
+
+        // Get the backup key as a single key
+        let backup_key_ed25519 = ed25519::new_unvalidated_public_key_from_bytes(cur_backup_key);
+        let backup_key_as_single_key = single_key::from_ed25519_public_key_unvalidated(&backup_key_ed25519);
+
+        // Construct the multi key public key which should be the current public key of the account
+        let account_public_key_bytes = vector[];
+        account_public_key_bytes.push_back(0x02); // Number of keys in the multi key
+        account_public_key_bytes.append(keyless_public_key);  // Add the current key which is a single key
+        account_public_key_bytes.append(single_key::unvalidated_public_key_to_bytes(&backup_key_as_single_key));
+        account_public_key_bytes.push_back(0x01); // Signatures required
+
+        // Check that constructed multi key public key is the current public key of the account by comparing
+        // its authentication key to the account's authentication key
+        let auth_key = get_authentication_key_from_scheme_and_public_key_bytes(MULTI_KEY_SCHEME, account_public_key_bytes);
+        assert!(
+            auth_key == account_resource.authentication_key,
+            error::invalid_argument(EWRONG_CURRENT_PUBLIC_KEY)
+        );
+
+        // Check that the keyless public key is the original public key of the account by comparing
+        // its authentication key to the account address.
+        let auth_key_for_original_public_key = get_authentication_key_from_scheme_and_public_key_bytes(SINGLE_KEY_SCHEME, keyless_public_key);
+        assert!(
+            bcs::to_bytes(&addr) == auth_key_for_original_public_key,
+            error::invalid_argument(ENOT_THE_ORIGINAL_PUBLIC_KEY)
+        );
+
+        let curr_auth_key_as_address = from_bcs::to_address(account_resource.authentication_key);
+        let challenge = RotationProofChallenge {
+            sequence_number: account_resource.sequence_number,
+            originator: addr,
+            current_auth_key: curr_auth_key_as_address,
+            new_public_key: new_backup_key,
+        };
+
+        // Assert the challenge signed by the new backup key is valid
+        assert_valid_rotation_proof_signature_and_get_auth_key(
+            ED25519_SCHEME,
+            new_backup_key,
+            new_backup_key_proof,
+            &challenge
+        );
+
+        let new_backup_key_ed25519 = ed25519::new_unvalidated_public_key_from_bytes(new_backup_key);
+        let new_backup_key_as_single_key = single_key::from_ed25519_public_key_unvalidated(&new_backup_key_ed25519);
+
+        // Construct the new multi key public key which contains the keyless public key and the new backup key
```

Comment:
Lots of duplicate logic to construct multikey (twice in this function and once in the add function). Consider refactoring into common function